dali/python/nvidia/dali/plugin/pytorch/proxy/__init__.py

```diff
@@ -0,0 +1,329 @@
+# Copyright (c) 2024, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import torch
+import torch.multiprocessing as mp
+from torch.utils import data as torchdata
+from torch.utils.data._utils.collate import default_collate_fn_map
+from nvidia.dali.backend import TensorGPU
+from nvidia.dali import Pipeline
+from nvidia.dali.external_source import ExternalSource
+import threading
+from queue import Empty
+from nvidia.dali.plugin.pytorch.torch_utils import to_torch_type, feed_ndarray
+import tree
+import warnings
+
+
+def _external_source_node_names(pipeline):
+    if not pipeline._py_graph_built:
+        pipeline._build_graph()
+    input_node_names = []
+    for op in pipeline._ops:
+        if isinstance(op._op, ExternalSource):
+            input_node_names.append(op.name)
+    return input_node_names
+
+
+class DALIPipelineOutputRef:
+    """
+    Placeholder for a pipeline output reference, after the iteration has been scheduled to DALI.
+    """
+
+    def __init__(self, info):
+        self.info = info
+
+
+class PipelineRunRef:
+    """
+    Placeholder for a pipeline run reference, which is returned by the data worker instead of
+    the actual data. The PyTorch worker returns this trivial object, only containing information
+    about this proxy instance and the input data to the pipeline. Later in the collate function,
+    we send the data for execution to DALI.
+    """
+
+    def __init__(self, proxy, inputs):
+        self.proxy = proxy
+        self.inputs = inputs
+        if len(inputs) != len(self.proxy.input_names):
+            raise RuntimeError(
+                f"Unexpected number of inputs. Expected: {self.input_names}, got: {inputs}"
+            )
+
+
+class _DALIProxy:
+    def __init__(self, input_names, send_q):
+        self.input_names = input_names
+        # Shared queue with the server
+        self.send_q = send_q
+        # Torch worker id, to be filled on first call to worker_id()
+        self._worker_id = None
+        # Iteration index for the current worker
+        self.data_idx = 0
+
+    @property
+    def worker_id(self):
+        """Getter for 'worker_id'"""
+        if self._worker_id is None:
+            worker_info = torchdata.get_worker_info()
+            self._worker_id = worker_info.id if worker_info else threading.get_ident()
+        return self._worker_id
+
+    def schedule_batch(self, inputs):
```

Comment:
I think this should be a private method - I know it's already a private class, but then again we expose it to the user through a public property - this seems fishy.